src/main/java/present/Search.java

```diff
@@ -15,6 +23,78 @@
    * @return the {@code n} nearest neighborhoods, ordered nearest to farthest
    */
   public static List<Neighborhood> near(Location location, int n) {
-    throw new UnsupportedOperationException("Please implement!");
+    Map<Neighborhood, Double> sortedNeighborhoodMap = getSortedNeighborhoodMapFromCache(location);
+    List<Neighborhood> nearestNeighborhoodList = new ArrayList<>();
+
+    int counter = 0;
+    for (Neighborhood neighborhood : sortedNeighborhoodMap.keySet()) {
+      nearestNeighborhoodList.add(neighborhood);
+
+      if (++counter >= n) {
+        break;
+      }
+    }
+
+      return nearestNeighborhoodList;
+  }
+
+  /**
+   * Get a sorted map from cache where the Key is a Neighborhood and the Value is the distance to it. If not found, compute all the distances, add it to cache and return it.
+   *
+   * @param location to consider when computing distances to all other locations
+   * @return sorted map of neighborhoods and the distance to them
+   */
+  private static Map<Neighborhood, Double> getSortedNeighborhoodMapFromCache(Location location) {
+    Map<Neighborhood, Double> distanceToNeighborhoodsSortedMap = NEAREST_NEIGHBORHOOD_CACHE.get(location);
+
+    if (distanceToNeighborhoodsSortedMap != null) {
+      return distanceToNeighborhoodsSortedMap;
+    }
+
+    distanceToNeighborhoodsSortedMap = getSortedNeighborhoodMap(location);
+
+    NEAREST_NEIGHBORHOOD_CACHE.put(location, distanceToNeighborhoodsSortedMap);
+
+    return distanceToNeighborhoodsSortedMap;
+  }
+
+  /**
+   *
+   * Get a sorted map getting a map of all neighborhoods and distance and then using Java 8 streams to sort it by value (distance) into a LinkedHashMap
+   *
+   * @param location to consider when computing distances to all other locations
+   * @return sorted map of neighborhoods and the distance to them
+   */
+  private static Map<Neighborhood, Double> getSortedNeighborhoodMap(Location location) {
+    Map<Neighborhood, Double> distanceToNeighborhoodsMap = getNeighborhoodMap(location);
+
+    return distanceToNeighborhoodsMap.entrySet().stream()
+            .sorted(Map.Entry.comparingByValue())
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue,
+                    (oldValue, newValue) -> oldValue, LinkedHashMap::new));
+  }
+
+  /**
+   *
+   * Get a map of all neighborhoods and the distance to them from current location
+   *
+   * @param location to consider when computing distances to all other locations
+   * @return map of neighborhoods and the distance to them
+   */
+  private static Map<Neighborhood, Double> getNeighborhoodMap(Location location) {
```

Comment:
A more descriptive name to consider: getDistancesByNeighborhood()

Reply:
Sure. Makes sense.